x/mongo/driver/topology/rtt_monitor_test.go

```diff
@@ -0,0 +1,212 @@
+				interval:       10 * time.Second,
+				wantSamplesLen: 30,
+			},
+			{
+				desc:           "min",
+				interval:       10 * time.Minute,
+				wantSamplesLen: 5,
+			},
+			{
+				desc:           "max",
+				interval:       1 * time.Millisecond,
+				wantSamplesLen: 500,
+			},
+		}
+		for _, tc := range cases {
+			t.Run(tc.desc, func(t *testing.T) {
+				rtt := newRTTMonitor(&rttConfig{
+					interval:     tc.interval,
+					minRTTWindow: 5 * time.Minute,
+				})
+				assert.Equal(t, tc.wantSamplesLen, len(rtt.samples), "expected samples length to match")
+			})
+		}
+	})
+
+	t.Run("can connect and disconnect repeatedly", func(t *testing.T) {
+		t.Parallel()
+
+		dialer := DialerFunc(func(_ context.Context, _, _ string) (net.Conn, error) {
+			reply := makeHelloReply()
+			return newSlowConn(10*time.Millisecond, reply[:4], reply[4:]), nil
+		})
+		rtt := newRTTMonitor(&rttConfig{
+			interval: 10 * time.Second,
+			createConnectionFn: func() (*connection, error) {
+				return newConnection("", WithDialer(func(Dialer) Dialer {
+					return dialer
+				}))
+			},
+			createOperationFn: func(conn driver.Connection) *operation.Hello {
+				return operation.NewHello().Deployment(driver.SingleConnectionDeployment{C: conn})
+			},
+		})
+		for i := 0; i < 100; i++ {
+			rtt.connect()
+			rtt.disconnect()
+		}
+	})
+
+	t.Run("works", func(t *testing.T) {
+		t.Parallel()
+
+		dialer := DialerFunc(func(_ context.Context, _, _ string) (net.Conn, error) {
+			reply := makeHelloReply()
+			return newSlowConn(10*time.Millisecond, reply[:4], reply[4:]), nil
+		})
+		rtt := newRTTMonitor(&rttConfig{
+			interval: 10 * time.Millisecond,
+			createConnectionFn: func() (*connection, error) {
+				return newConnection("", WithDialer(func(Dialer) Dialer {
+					return dialer
+				}))
+			},
+			createOperationFn: func(conn driver.Connection) *operation.Hello {
+				return operation.NewHello().Deployment(driver.SingleConnectionDeployment{C: conn})
+			},
+		})
+		rtt.connect()
+		defer rtt.disconnect()
+
+		// Wait up to a second for the avg and min RTTs to be greater than 0.
+		start := time.Now()
+		for time.Since(start) < 1*time.Second {
+			if rtt.getRTT() > 0 && rtt.getMinRTT() > 0 {
+				break
+			}
+		}
+
+		assert.Greater(t, rtt.getMinRTT().Nanoseconds(), int64(0), "expected getMinRTT() to return a positive duration")
+		assert.Greater(t, rtt.getRTT().Nanoseconds(), int64(0), "expected getRTT() to return a positive duration")
+	})
+}
+
+var _ net.Conn = &slowConn{}
+
+// slowConn is a net.Conn that returns a response after sleeping for a specified time.
+type slowConn struct {
+	duration  time.Duration
+	responses [][]byte
+	offset    int
+}
+
+func newSlowConn(duration time.Duration, responses ...[]byte) *slowConn {
+	return &slowConn{
+		duration:  duration,
+		responses: responses,
+	}
+}
+
+func (c *slowConn) Read(b []byte) (int, error) {
+	time.Sleep(c.duration)
+	response := c.responses[c.offset]
+	c.offset = (c.offset + 1) % len(c.responses)
```

Comment:
More specifically, it assumes that `cap(b) > len(c.response[c.offset])` because `copy` will stop at `min(len(dst), len(src))`. Fortunately that works because `connection.read()` allocates a destination buffer of exactly the size specified by the first 4 bytes of the server reply. I actually think this can be simplified to use a byte buffer instead of a slice of byte slices.